konnectivity-client/pkg/client/metrics/metrics.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	sharedmetrics "sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/shared/metrics"
+	"sigs.k8s.io/apiserver-network-proxy/konnectivity-client/proto/client"
+)
+
+const (
+	Namespace = "apiserver"
+	Subsystem = "konnectivity_client"
+)
+
+var (
+	// Metrics provides access to all client metrics. The client
+	// application is responsible for registering (via Metrics.RegisterMetrics).
+	Metrics = newMetrics()
+)
+
+// ClientMetrics includes all the metrics of the konnectivity-client.
+type ClientMetrics struct {
+	tConnections       *prometheus.GaugeVec
+	streamEvents       *prometheus.CounterVec
+	streamEventsErrors *prometheus.CounterVec
+	registerOnce       sync.Once
+}
+
+func newMetrics() *ClientMetrics {
+	tConnections := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: Namespace,
+			Subsystem: Subsystem,
+			Name:      "tunnel_connections",
+			Help:      "Current number of open tunnel connections.",
+		},
+		[]string{},
+	)
+	streamEvents := sharedmetrics.MakeStreamEventsTotalMetric(Namespace, Subsystem)
+	streamEventsErrors := sharedmetrics.MakeStreamEventsErrorsMetric(Namespace, Subsystem)
+	return &ClientMetrics{
+		tConnections:       tConnections,
+		streamEvents:       streamEvents,
+		streamEventsErrors: streamEventsErrors,
+	}
+}
+
+// RegisterMetrics registers all metrics, given a prometheus MustRegister function.
+func (c *ClientMetrics) RegisterMetrics(mustRegisterFn func(...prometheus.Collector)) {
```

Comment:
I would pass in a `prometheus.Registerer` instead.

Reply:
I couldn't find anything in legacyregistry that implements the Registerer interface. Maybe I'm missing something?

Reply:
https://github.com/kubernetes/kubernetes/blob/3e26e104bdf9d0dc3c4046d6350b93557c67f3f4/staging/src/k8s.io/component-base/metrics/wrappers.go#L133-L140

Reply:
@logicalhan Thanks for the quick turnaround!

Just to be cautious of dependencies I expose both old and new way to register in this PR.